crates/codegen/parser/runtime/src/napi/templates/ast_selectors.rs.jinja2

```diff
@@ -0,0 +1,342 @@
+    impl Selector {
+        fn {{ separated.name | snake_case }}(&mut self) -> Result<Vec<Vec<JsObject>>> {
+            let mut separated = vec![];
+            let mut separators = vec![];
+
+            separated.push(self.select(|node| {
+                {%- if separated.is_terminal -%}
+                    node.is_token_with_kind(TokenKind::{{ separated.reference }})
+                {%- else -%}
+                    node.is_rule_with_kind(RuleKind::{{ separated.reference }})
+                {%- endif -%}
+            })?);
+
+            while let Some(separator) = self.try_select(|node| node.is_token_with_kind(TokenKind::{{ separated.separator }}))? {
+                separators.push(separator);
+
+                separated.push(self.select(|node| {
+                    {%- if separated.is_terminal -%}
+                        node.is_token_with_kind(TokenKind::{{ separated.reference }})
+                    {%- else -%}
+                        node.is_rule_with_kind(RuleKind::{{ separated.reference }})
+                    {%- endif -%}
+                })?);
+            }
+
+            Ok(vec![separated, separators])
+        }
+    }
+{% endfor %}
+
+//
+// Common:
+//
+
+struct Selector {
+    env: Env,
+    node: Rc<RustRuleNode>,
+    index: usize,
+}
+
+impl Selector {
+    fn new(node: &RuleNode, env: Env) -> Self {
+        Self {
+            env,
+            node: node.0.clone(),
+            index: 0,
+        }
+    }
+
+    fn select(&mut self, filter: impl FnOnce(&RustNode) -> bool) -> Result<JsObject> {
+        match self.try_select(filter)? {
+            Some(node) => Ok(node),
+            None => Error::MissingChild(self.index).into(),
+        }
+    }
+
+    fn try_select(&mut self, filter: impl FnOnce(&RustNode) -> bool) -> Result<Option<JsObject>> {
+        while let Some(child) = self.node.children.get(self.index) {
+            match child {
+                RustNamedNode {
+                    name: _,
+                    node: RustNode::Rule(rule),
+                } if matches!(
+                    rule.kind,
+                    RuleKind::LeadingTrivia | RuleKind::TrailingTrivia | RuleKind::EndOfFileTrivia
+                ) => {
+                    // skip trivia, since it's not part of the AST
+                    self.index += 1;
+                    continue;
+                }
+                RustNamedNode {
+                    name: _,
+                    node: RustNode::Token(token),
+                } if matches!(token.kind, TokenKind::SKIPPED) => {
+                    return Error::SkippedToken(self.index).into();
+                }
+                node if filter(node) => {
+                    self.index += 1;
+                    return Ok(Some(node.to_js(&self.env)));
+                }
+                _ => {
+                    break;
+                },
+            }
+        }
+
+        Ok(None)
+    }
+
+    fn finalize(&mut self) -> Result<()> {
```

Comment:
nit: If I understand the intent correctly, this should probably consume the type with `self` to better express intent/disallow calling twice